### ISSUE TITLE
Bringing package scripts into the fold

### ISSUE THREAD

**Issue by slproweb** [openssl/openssl#20865]
I've been building the semi-official binary releases of OpenSSL for Windows for many years.  Had a good chat with Tam, Anton, and Hugo this morning and want to start the process of getting the scripts I currently use cleaned up and imported "somewhere" under the formal purview of the main OpenSSL project.  This issue exists to make sure the process is done with positive public input and transparency.

Currently, I use a hodgepodge of technologies to produce the binaries in 3 to 4 phases (depending on how you want to view the process), ~~including one old Borland C++ binary from some ancient version of that compiler~~.  There are a number of hardcoded paths, which will make it a bit difficult to replicate the setup I use in other environments.  So I believe a progressive approach will be best rather than trying to do everything that needs to be done all at once.

The first thing to do will be to decide where various things need to live.  Should the scripts for Windows live off of the 'ms' subdirectory or live in its own repo or something else entirely different?  A separate repo seems like the best option to me to afford less confusion and those with package script access aren't necessarily the same people as those with main repo access.  If a repo is the preferred route, let's get that setup as well as whatever relevant access, if any - a pull request/merge approach is fine with me as well.  This whole process will likely take a while, so I don't think there's a rush to get this done in time for the next version tag as long as it happens and active progress is made.

I'm happy to do the work involved and look forward to feedback on this endeavor.

**Comment by slproweb** [comment]
> Are the scripts version specific? i.e. do you have 3.0 scripts and separate 1.1.1 scripts? Or do the same scripts cover multiple versions?

Some of both.  The InnoSetup/Wix installer scripts themselves are version specific.  That is partly due to changes to OpenSSL that happen over time but also maintaining the same version of Visual Studio across version releases.  For example, 1.1.1 is built against VS 2017 while 3.0 and 3.1 are built against VS 2019.  (3.1 is still VS 2019 because when I went to install VS 2022, the installer for VS failed to install several important components.  I wasn't confident on the stability of that install, so I ended up going back to VS 2019.)

There are several scripts I use that are more generic-ish and not necessarily version specific.  I still end up tweaking those a little here and there to deal with new release issues but they are largely stable.

**Comment by levitte** [comment]
I would put the packaging stuff in a separate repository, for now.  It's possible that we can integrate it into the main source alter on, but I believe that will take some scrutiny.

**Comment by slproweb** [comment]
Sounds good to me.  I've already begun work on this project.  I'm taking what I've currently got and I'm reworking it to:  Clean up folder structures, eliminate hardcoded paths, and make a simplified, unified tool that will hopefully be more generic and repeatable.  Fewer things that can go wrong is generally better for everyone.

It'll be a few more days before I have anything ready.  I'll start posting some updates once I get the first working build.

**Comment by hlandau** [comment]
We discussed this and the feeling is we should create a repository called `openssl/installer` with a subdirectory for Windows (could contain non-Windows installers in future, but just Windows for now).

We'll get this repo created and then you can raise a PR against it. And again, thanks for the effort.

**Comment by slproweb** [comment]
I've got the first set of installers building successfully.  The unified script can only build OpenSSL 3.1 for x86 (VC-WIN32) at the moment but expanding that to other base versions and architectures should happen soon.

For now I'm just aiming for parity/equivalence to what the existing scripts do with a few minor tweaks/improvements here and there (e.g. reducing 8 builds to 6).

**Comment by slproweb** [comment]
Okay, over the past few days, I got all architectures building for 1.1.1 (x86 and x64), 3.0 (x86, x64, and arm64), and 3.1 (x86, x64, and arm64).  I also added ZIP file generation for two portable edition flavors.

Since I don't see an `openssl/installer` repository, I went ahead and created a temporary repo:

https://github.com/slproweb/openssl-packagers/tree/main/ms-windows

Addressing the elephant in the room up front:  Why is the unified build tool written in PHP?  Mostly because I know PHP and also because the existing scripts are/were written in PHP (and .bat files).  It was faster (it only took one week) and easier to simply port all of the existing code over into the new tool and clean up the code than to try to rewrite it all at the same time in a language I don't know and would have to go find a beginner's tutorial for (i.e. I'm not all that familiar with Perl beyond running it for the Configure step).  PHP also happens to run just fine on the command line.  At any rate, getting something working first was more important.

With that out of the way, feel free to analyze the code and templates, provide feedback, ask questions, etc.  I tried to comment and document everything fairly liberally and organized the code and resources to hopefully make it easier to follow along with and understand.

The 'deps' and 'profiles' portions, the base version JSON files, and some of the automation bits are, more or less, the only brand new code and resources.  Everything else is largely untouched and just reworked slightly to fit into the unified tool.  I did notice that 3.0 and 3.1 have virtually identical Inno Setup and Wix toolset scripts and JSON after transforming them into templated variants, which means there could be a way to simplify those even further.

**Comment by mattcaswell** [comment]
Looks very nice!

I've read through the README and scanned through various of the template files to get a feel for it.

I don't think PHP is a big concern. It wouldn't be acceptable for the main source repo because that code has to be able to run *everywhere* and on all sorts of platforms and environments. But for *this* code we only need it to run in certain well defined environments and its not a big deal to add additional build-time dependencies for that.

I am a noob as far as windows installers are concerned so I can only provide limited feedback - but I spotted some interesting things in templates/3.1/info.json:

````
				"configure_target": "VC-WIN32",
				"configure_extra": "-D_USE_32BIT_TIME_T -D_WINSOCK_DEPRECATED_NO_WARNINGS -D_WIN32_WINNT=0x0501",
````
and

````
				"configure_target": "VC-WIN64A",
				"configure_extra": "-D_WINSOCK_DEPRECATED_NO_WARNINGS -D_WIN32_WINNT=0x0502",
````

`0x0501` corresponds to Windows XP I think, and `0x0502` is Windows Server 2003. I wonder whether those are the correct versions to be targeting. OpenSSL itself is supposed to be able to support anything from Windows XP onwards - but whether we want to support that far back in a binary is another question. IIRC there are various optimisations in the code if we are allowed to assume a later version.

I'm also interested in why the other defines are needed.

Anyway, it looks very good so far. It will be easier to review once we have a repository set up and we can create PRs against it.

**Comment by slproweb** [comment]
Here's the full build section from the JSON:

```
			"build": {
				"configure_target": "VC-WIN32",
				"configure_extra": "-D_USE_32BIT_TIME_T -D_WINSOCK_DEPRECATED_NO_WARNINGS -D_WIN32_WINNT=0x0501",
				"patches": {
					"/Zi ": "/Z7 ",
					"/subsystem:console ": "/subsystem:console,5.01 ",
					"CNF_LDFLAGS=": "CNF_LDFLAGS=/subsystem:console,5.01 "
				}
```

Above was extracted from:

https://github.com/slproweb/openssl-packagers/blob/main/ms-windows/templates/3.1/info.json

The defines were pulled from the existing build scripts.  They are various tweaks I made to the Makefile over the years to address various concerns that came up.  Some of those concerns may or may not still be valid but I'll try to address them here from what I recall:

The `_USE_32BIT_TIME_T` define originally came from:  http://groups.google.com/group/mailing.openssl.users/browse_thread/thread/006b8e8b8c8463d0/447180a3a6c167bf?lnk=raot

The `_WINSOCK_DEPRECATED_NO_WARNINGS` define has been in there basically forever.  I think at one point the deprecated Winsock warnings triggered a build failure.  Adding the define made the problem go away.

The `_WIN32_WINNT` version define is used by the Windows SDK to prevent newer APIs from being used.  It doesn't guarantee that won't happen though.  0x0502 is the first version of Windows with a 64-bit build of the OS.  I'm pretty sure that was Windows XP Professional, x64 edition, which never really saw broad deployment, but it did exist.

The `/Zi` to `/Z7` patch deals with linker issues when mixing versions of the Visual C++ compiler/linker.  The generated libraries may work with one version of the compiler/linker but not others.  I don't recall the exact problem but someone suggested that as the fix, I researched the options to make sure it seemed sane, and it seemed to work because the problems vanished.

The two `/subsystem:console,5.01` patch lines force the final linked EXEs/DLLs to declare themselves to be Windows XP compatible.  That doesn't mean the executable itself will load and run (e.g. DLL loader failures) but at least it won't be due to an artificial limitation within the EXE itself.

Just to point out that this is a nearly 1:1 port of my existing code - weird build warts and all - with some tweaks to cater better to the project's needs.  You don't _have_ to use those patches or you could backport some or all of the patches into the mainline or do something entirely different.  This was a first pass to get everything into one place as it currently exists.

```
			// Nothing is perfect.  Patch the makefile.
			$data = file_get_contents($pathmap["[[SOURCE_TEMP_DIR]]"] . "/makefile");

			foreach ($archinfo["build"]["patches"] as $key => $val)
			{
				$data = str_replace($key, $val, $data);
			}

			// The 'no-shared' build should only generate statically linked libraries and executables.
			if (strpos($configuretype, "no-shared") !== false)
			{
				$data = str_replace("MODULES=providers\\legacy.dll", "MODULES=", $data);
				$data = str_replace("MODULEPDBS=providers\\legacy.pdb", "MODULEPDBS=", $data);
			}

			// Fix the build type.
			$data = str_replace(" /MTd ", " " . $buildtype . " ", $data);
			$data = str_replace(" /MTd\r\n", " " . $buildtype . "\r\n", $data);
			$data = str_replace(" /MT ", " " . $buildtype . " ", $data);
			$data = str_replace(" /MT\r\n", " " . $buildtype . "\r\n", $data);

			$data = str_replace(" /MDd ", " " . $buildtype . " ", $data);
			$data = str_replace(" /MDd\r\n", " " . $buildtype . "\r\n", $data);
			$data = str_replace(" /MD ", " " . $buildtype . " ", $data);
			$data = str_replace(" /MD\r\n", " " . $buildtype . "\r\n", $data);

			file_put_contents($pathmap["[[SOURCE_TEMP_DIR]]"] . "/makefile", $data);
```

Above was extracted from:

https://github.com/slproweb/openssl-packagers/blob/main/ms-windows/build-tools.php#L1209

The makefile itself is patched to replace the build type with the desired build type.  Those linking their applications need support for all four main build types (MD, MDd, MT, MTd) to avoid linker errors.  shared vs. no-shared mostly gets there but there's no way to directly declare the build type.  There are some problems with software built with MT and MTd but some people want and use them anyway.  Since the DLLs export C library functions, there's not a whole lot of concern over mixing MD linked DLLs against MT/MTd libraries, especially when using DLL local resources like the memory isolated BIO API.  Mixing VC++ runtimes is not great but multiple runtimes can be loaded just fine in memory in the same process space as long as one runtime doesn't try to free another runtime's allocated memory or resources.

As a side effect of the monkey patching that happens, the legacy provider DLL fails to build altogether in the no-shared builds for whatever reason.  That's why no-shared builds neutralize those two lines in the makefile patching above.  I don't think that the failure to build nor successfully building the legacy provider DLL matters here because the intent was to produce two statically linked variants of the main openssl.exe and the legacy provider DLL is successfully compiled and linked in other builds.  Fully statically linked OpenSSL is a bit different than dynamically linked anyway.

The ability to post-patch the makefile can also make it possible to apply fixes for release level bugs where something unexpected crops up at the last minute for a specific architecture.  3.0.8 had that issue with Windows ARM64 builds, albeit that was a source code problem rather than a makefile problem, but you get the idea.

**Comment by levitte** [comment]
Er...  looking into the support subdirectory.  Who is CubicleSoft?

**Comment by slproweb** [comment]
That's also me.  Second business entity.  Those libraries are open source:

https://github.com/cubiclesoft/php-misc
https://github.com/cubiclesoft/ultimate-web-scraper

**Comment by levitte** [comment]
openssl/installer now exists, feel free to submit your stuff as a PR.

**Comment by slproweb** [comment]
Unless I am missing something, I don't think submitting a PR against the new repository is currently possible.  Specifically, the GitHub web interface isn't letting me fork the repo:  "Cannot fork because repository is empty."

Based on some searching Google/Stack Overflow, I'm pretty sure there has to be at least one branch on a git repo to be able to fork and submit PRs.

**Comment by levitte** [comment]
Argh.  I always forget that bit.
I've pushed an embryo for a `.gitignore` file, so now there is a `main` branch.

**Comment by nhorman** [comment]
Marking as inactive.  will be closed at the end of 3.4 dev cycle

**Comment by paulidale** [comment]
Is there still going to be a pre-built Windows binary available from the project?

**Comment by slproweb** [comment]
I believe that will be the case.  It looks like Neil has taken over the installer and simplified it quite a bit.  I just left a note on the unfinished pull request that I had originally opened:

https://github.com/openssl/installer/pull/2#issuecomment-2153724731

**Comment by nhorman** [comment]
@paulidale @slproweb yes, our intent is to provide binary installers.  Currently the plan is to do this in two parts:

1) Create the binary installer and make openssl compatible with it, tracked here: https://github.com/openssl/project/issues/643

That epic tracks the creation of the installer, and the updates needed to make openssl compatible with it (most notably the creation of registry keys to allow what were previously build time constants (like OPENSSLDIR) to be install time selectable on a per-installation basis.  This is a 3.4 deliverable

2) Obtain a signing certificate from a vendor to sign the installer, create a process to build and sign the installer as part of our release process, and add the installer binary as a release artifact.  The build process (minus the signing) is available in 3.4 as a manually triggered ci workflow in the installer repo now, but the rest of this will be a 3.5 deliverable.
